### PR TITLE
Token provider authentication

### PR DESCRIPTION
**Pull request by ohadbitt**
#### Pull Request Description
---
**Features:**
- Added a new option ("tokenProvider") to provide class path to a Token provider class that extends callable[String] - the java client is created with an instance of the class.

### CODE REVIEW

File: connector/src/main/scala/com/microsoft/kusto/spark/authentication/KustoAuthentication.scala [hunk 1]
```diff
@@ -61,3 +63,12 @@ case class KustoAccessTokenAuthentication(token: String) extends KustoAuthentica
   override def hashCode(): Int = token.hashCode
 }
 
+case class KustoTokenProviderAuthentication(tokenProvider: Callable[String]) extends KustoAuthentication {
+  def canEqual(that: Any) : Boolean = that.isInstanceOf[KustoAccessTokenAuthentication]
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/authentication/KustoAuthentication.scala:67]:
```suggestion
  def canEqual(that: Any) : Boolean = that.isInstanceOf[KustoTokenProviderAuthentication]
```

File: connector/src/main/scala/com/microsoft/kusto/spark/common/KustoOptions.scala [hunk 2]
```diff
@@ -29,6 +29,10 @@ trait KustoOptions {
   // AAD application certificate password
   val KUSTO_AAD_APP_CERTIFICATE_PASSWORD: String = newOption("kutoAadAppCertPassword")
 
+  // Coordinates of a class that extends callable<String> with Serializeable and returns a token.
+  // The provider will be called for every request to the kusto service
+  val KUSTO_TOKEN_PROVIDER: String = newOption("tokenProvider")
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/common/KustoOptions.scala:34]:
```suggestion
  val KUSTO_TOKEN_PROVIDER_CALLBACK: String = newOption("tokenProviderCallback")
```

This is a callback, right?

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 3]
```diff
@@ -113,6 +113,8 @@ object KustoDataSourceUtils {
     val applicationKey = parameters.getOrElse(KustoSourceOptions.KUSTO_AAD_APP_SECRET, "")
     val applicationCertPath = parameters.getOrElse(KustoSourceOptions.KUSTO_AAD_APP_CERTIFICATE_PATH, "")
     val applicationCertPassword = parameters.getOrElse(KustoSourceOptions.KUSTO_AAD_APP_CERTIFICATE_PASSWORD, "")
+    val tokenProviderCoordinates = parameters.getOrElse(KustoSourceOptions.KUSTO_TOKEN_PROVIDER, "")
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:116]: Coordinates? did you mean callback?

```suggestion
    val tokenProviderCallback = parameters.getOrElse(KustoSourceOptions.KUSTO_TOKEN_PROVIDER, "")
```

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 4]
```diff
@@ -138,12 +140,20 @@ object KustoDataSourceUtils {
...
-    else if ( {
+    } else if ( {
       accessToken = parameters.getOrElse(KustoSourceOptions.KUSTO_ACCESS_TOKEN, "")
       !accessToken.isEmpty
     }) {
       authentication = KustoAccessTokenAuthentication(accessToken)
+    }  else if ( {
+      !tokenProviderCoordinates.isEmpty
+    }) {
+      val provider1 = Thread.currentThread().getContextClassLoader
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:151]:
```suggestion
      val classLoader= Thread.currentThread().getContextClassLoader
```

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 5]
```diff
@@ -138,12 +140,20 @@ object KustoDataSourceUtils {
...
         authentication = AadApplicationCertificateAuthentication(applicationId, applicationCertPath, applicationCertPassword);
       }
-    }
-    else if ( {
+    } else if ( {
       accessToken = parameters.getOrElse(KustoSourceOptions.KUSTO_ACCESS_TOKEN, "")
       !accessToken.isEmpty
     }) {
       authentication = KustoAccessTokenAuthentication(accessToken)
+    }  else if ( {
```
Comment (noam-shapira) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:148]: no need for  {

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 6]
```diff
@@ -138,12 +140,20 @@ object KustoDataSourceUtils {
       }else if(!applicationCertPath.isEmpty){
         authentication = AadApplicationCertificateAuthentication(applicationId, applicationCertPath, applicationCertPassword);
       }
-    }
-    else if ( {
+    } else if ( {
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:143]: This entire section is very messy.
I suggest to rewrite it and split into three sections:
1. Read all the options into local variables
2. Check conflicting options
3. Initialize the authentication handle properly

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 7]
```diff
@@ -138,12 +140,20 @@ object KustoDataSourceUtils {
...
       accessToken = parameters.getOrElse(KustoSourceOptions.KUSTO_ACCESS_TOKEN, "")
       !accessToken.isEmpty
     }) {
       authentication = KustoAccessTokenAuthentication(accessToken)
+    }  else if ( {
+      !tokenProviderCoordinates.isEmpty
+    }) {
+      val provider1 = Thread.currentThread().getContextClassLoader
+      val c1 = provider1.loadClass(tokenProviderCoordinates).newInstance()
+      val tokenProvider = c1.asInstanceOf[Callable[String]]
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:153]:
```suggestion
      val tokenProviderCallback = c1.asInstanceOf[Callable[String]]
```

**Review by vladikbr — CHANGES_REQUESTED** [review]
See comments

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala (hunk 3 above)
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:116]: i meant the class path to the tokenProvider
i.e 
com.microsoft.kusto.spark.authentication.TokenProvider

File: connector/src/main/scala/com/microsoft/kusto/spark/common/KustoOptions.scala (hunk 2 above)
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/common/KustoOptions.scala:34]: Yea
Its actually class that extents callable

Comment (vladikbr) [reply connector/src/main/scala/com/microsoft/kusto/spark/common/KustoOptions.scala:34]: Ok, then it's KUSTO_TOKEN_PROVIDER_CALLBACK_CLASSPATH

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 8]
```diff
@@ -113,38 +113,59 @@ object KustoDataSourceUtils {
...
     }
 
+    // Look for conflicts
+    var numberOfAuthenticationMethods = 0
+    if (!applicationId.isEmpty) numberOfAuthenticationMethods += 1
+    if (!accessToken.isEmpty) numberOfAuthenticationMethods += 1
+    if (!tokenProviderCoordinates.isEmpty) numberOfAuthenticationMethods += 1
+    if (!keyVaultUri.isEmpty) numberOfAuthenticationMethods += 1
+    if(numberOfAuthenticationMethods > 1){
+      throw new IllegalArgumentException("More than one authentication methods were provided. Failing.")
```
Comment (vladikbr) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala:146]:
```suggestion
      throw new IllegalArgumentException("Conflicting authentication settings specified.")
```

**Review by vladikbr — CHANGES_REQUESTED** [review]
final touch...

**Review by vladikbr — APPROVED** [review]

**Pull request merged**